blaze-server/src/main/scala/org/http4s/blaze/server/Http1ServerStage.scala

```diff
@@ -229,15 +229,13 @@ private[blaze] class Http1ServerStage[F[_]](
               F.runCancelable(action) {
                 case Right(()) => IO.unit
                 case Left(t) =>
-                  IO(logger.error(t)(s"Error running request: $req")).attempt *> IO(
-                    closeConnection()
-                  )
+                  IO(logger.error(t)(s"Error running request: $req")).attempt *>
+                    IO.delay(cancelToken.set(None)) *>
+                    IO(closeConnection())
               }.unsafeRunSync()
             )
 
-            parser.synchronized {
```

Comment:
I actually don't get why we need that synchronization here. If anyone can explain this, I'd be very grateful.

Reply:
`cancel()` is also synchronized on `parser`, so it linearizes those.  I'm not awake enough yet to be 100% sure to what end.

Reply:
@RafalSumislawski moved the running of the action out of this block in #4761, but left the synchronization.  I'd be interested to hear if he can think of any reason it needs to remain.

Reply:
We need this synchronised for it's Java Memory Model effects. We need a `cancelToken = theCancelToken` -happens-before-> `cancelToken.foreach {...}`. The syncronized provides that thanks to the fact that the `cancelToken.foreach {...}` sits in the `cancel` methood which is called only from `parser.syncrionized{...}` block inside `stageShutdown`. 

`volatile` will do the job as well.